File: 238-product-of-array-except-self.py
# look at js file for reasoning
# iterating backwards in python can be done by using -1 in 3rd arg of range

class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        res = list(nums)
        forward = 1
        backward = 1

        for i in range(len(nums)-1, -1, -1):
            res[i] = backward
            backward *= nums[i]

        for j in range(len(nums)):
            res[j] *= forward
            forward *= nums[j]

        return res

File: 238-product-of-array-except-self.js
/*
given int array nums, return array where res[i] is the product of elements in nums excep itself
*/

/*
1.) this can be done with 2 loops (keeps time complexity O(n)), one iterating forward across nums and the other iterating backwards across nums
2.) will also need two variables to hold what is being multiplied in each loop (one for each loop) (start both at 1)
3.) in the backward loop, fill result array with the mult variable at each corresponding index
4.) in the forward loop, multiply corresponding indices of result by forward loops mult variable
5.) in both loops, the mult variable is updated on each iteration by multiplying by that index of nums
*/

var productExceptSelf = function(nums) {
    const res = [];
    let leftMult = 1;
    let rightMult = 1;
    for (let i=nums.length - 1; i >= 0; i--) {
        res[i] = rightMult;
        rightMult *= nums[i];
    }
    for (let j=0; j < nums.length; j++) {
        res[j] *= leftMult;
        leftMult *= nums[j];
    }
    return res;
};
